libs/dyn/pattern_trie.go

```diff
@@ -0,0 +1,182 @@
+package dyn
+
+import "fmt"
+
+// PatternTrie is a trie data structure for storing and querying patterns.
+// It supports both exact matches and wildcard matches. You can insert [Pattern]s
+// into the trie and then query it to see if a given [Path] matches any of the
+// patterns.
+type PatternTrie struct {
+	root *trieNode
+}
+
+// trieNode represents a node in the pattern trie.
+// Note that it can only be one of anyKey, anyIndex, collection of pathKeys, or collection of pathIndexes.
+type trieNode struct {
+	// If set this indicates the trie node is an anyKey node.
```

Comment:
Would be good to annotate that `AnyKey` and `AnyIndex` are names of existing structs in the `dyn` package. Otherwise, I got confused by this naming (expecting to see a generic Trie implementation here)

Reply:
Good callout. Added.